File: utils/pageupdater/pages.py
import os
import hashlib
from typing import Dict
# Front Matter reader
import frontmatter

def get_metadata(file_path):
    file = frontmatter.load(file_path)
    return file.metadata

def get_sha256_hash(file_path):
    # Open the file in binary mode
    with open(file_path, "rb") as f:
        # Read the contents of the file
        content = f.read()
        # Create a sha256 hash object
        sha256_hash = hashlib.sha256()
        # Update the hash object with the file content
        sha256_hash.update(content)
        # Get the hexadecimal representation of the hash
        hex_dig = sha256_hash.hexdigest()
    return hex_dig

def get_char_count(file_path):
    # Open the file in read mode
    with open(file_path, "r") as f:
        # Read the contents of the file
        content = f.read()
        # Get the character count of the content
        char_count = len(content)
    return char_count

def get_word_count(file_path):
    # Open the file in read mode
    with open(file_path, "r") as f:
        # Read the contents of the file
        content = f.read()
        # Get the word count of the content
        word_count = len(content.split())
    return word_count

# Dict of pages containing metadata (Front Matter), sha256 hash of the page content,
# page character count, page word count, and path
def get_pages_info(search_directory, root_directory):
    # Scan the search directory for .md files
    # When seeing a directory, recursively call this function

    # Initialize the dictionary
    page_info = {}

    current_directory = root_directory + search_directory

    # Get the list of files and directories in the search directory
    files = os.listdir(root_directory + search_directory)

    # Iterate over the files and directories
    for file in files:
        # Get the full path of the file
        full_path = os.path.join(current_directory, file)
        local_path = full_path.replace(root_directory, "")

        # If the file is a directory, recursively call this function
        if os.path.isdir(full_path):
            page_info.update(get_page_infos(search_directory + "/" + file, root_directory))
        # If the file is a markdown file, get the metadata
        elif file.endswith(".md"):
            # Get the metadata
            metadata = get_metadata(full_path)

            # Get the sha256 hash of the content
            sha256_hash = get_sha256_hash(full_path)

            # Get the character count of the content
            char_count = get_char_count(full_path)

            # Get the word count of the content
            word_count = get_word_count(full_path)

            # Add the metadata, sha256 hash, character count, word count, and path to the dictionary
            page_info[full_path] = {
                "local_path": local_path,
                "metadata": metadata,
                "hash": sha256_hash,
                "char_count": char_count,
                "word_count": word_count,
                "path": full_path,
            }
    
    return page_info

File: utils/page_updater/pages.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as crypto from 'node:crypto';
import fm, { type FrontMatterResult } from 'front-matter';

export interface PageLocation {
    title: string;
    description: string;
    tags: string[];
    map: string;
    root: string;
    show: boolean;
}

export interface PageInfoMetdata {
    title?: string;
    description?: string;
    date?: string;
    tags?: string[];
    background?: string;
    next?: string ;
    previous?: string;
}

// Type for metadata and page info
export interface PageInfo {
    local_path: string;
    absolute_path: string;
    metadata: PageInfoMetdata;
    hash: string;
    char_count: number;
    word_count: number;
    path: string;
}

export interface PageInfoCategory {
    posts: PageInfo[];
    title: string;
    description: string;
    tags: string[];
}

// Function to get metadata from a file
function getMetadata(filePath: string): PageInfoMetdata {
    const fileContent = fs.readFileSync(filePath).toString();
    return fm(fileContent).attributes as PageInfoMetdata
}

// Function to get SHA-256 hash of a file
function getSha256Hash(filePath: string): string {
    const fileContent = fs.readFileSync(filePath);
    const hash = crypto.createHash('sha256');
    hash.update(fileContent.toString());
    return hash.digest('hex');
}

// Function to get character count of a file
function getCharCount(filePath: string): number {
    const fileContent = fs.readFileSync(filePath, 'utf8');
    return fileContent.length;
}

// Function to get word count of a file
function getWordCount(filePath: string): number {
    const fileContent = fs.readFileSync(filePath, 'utf8');
    return fileContent.split(/\s+/).length;
}

// Function to get pages info
export function getPagesInfo(searchDirectory: string, pageLocation: PageLocation): Record<string, PageInfo> {
    const pageInfo: Record<string, PageInfo> = {};
    const currentDirectory = path.join(pageLocation.root, searchDirectory);
    const files = fs.readdirSync(currentDirectory);
    console.log(files);

    files.forEach((file) => {
        var fullPath = path.join(currentDirectory, file);
        var localPath = fullPath.replace(pageLocation.root, pageLocation.map);
        console.log(fullPath);
        console.log(localPath);
        if (fs.lstatSync(fullPath).isDirectory()) {
            Object.assign(pageInfo, getPagesInfo(path.join(searchDirectory, file), pageLocation));
        } else if (file.endsWith('.md')) {

            const metadata = getMetadata(fullPath);
            const sha256Hash = getSha256Hash(fullPath);
            const charCount = getCharCount(fullPath);
            const wordCount = getWordCount(fullPath);

            // Remove the .md extension
            localPath = localPath.replace('.md', '');
            fullPath = fullPath.replace('.md', '');

            pageInfo[fullPath] = {
                local_path: localPath,
                absolute_path: fullPath,
                metadata: metadata,
                hash: sha256Hash,
                char_count: charCount,
                word_count: wordCount,
                path: fullPath
            };
        }
    });

    return pageInfo;
}
